src/generic_core/uproxy_core.ts

```diff
@@ -44,19 +44,59 @@ loggingController.setDefaultFilter(
 var portControl = globals.portControl;
 
 // Prefix for freedomjs modules which interface with cloud computing providers.
-const CLOUD_PROVIDER_MODULE_PREFIX: string = 'CLOUDPROVIDER-';
+const CLOUD_PROVIDER_MODULE_NAME_PREFIX: string = 'CLOUDPROVIDER-';
+const CLOUD_PROVIDER_NAME = 'digitalocean';
+const CLOUD_PROVIDER_MODULE_NAME = CLOUD_PROVIDER_MODULE_NAME_PREFIX + CLOUD_PROVIDER_NAME;
+const CLOUD_INSTALLER_MODULE_NAME = 'cloudinstall';
 
 const getCloudProviderNames = (): string[] => {
   let results: string[] = [];
   for (var dependency in freedom) {
     if (freedom.hasOwnProperty(dependency) &&
-        dependency.indexOf(CLOUD_PROVIDER_MODULE_PREFIX) === 0) {
-      results.push(dependency.substr(CLOUD_PROVIDER_MODULE_PREFIX.length));
+        dependency.indexOf(CLOUD_PROVIDER_MODULE_NAME_PREFIX) === 0) {
+      results.push(dependency.substr(CLOUD_PROVIDER_MODULE_NAME_PREFIX.length));
     }
   }
   return results;
 };
 
+// This is the name recommended by the blog post.
+const CLOUD_DROPLET_NAME = 'uproxy-cloud-server';
+
+// Percentage of cloud install progress devoted to deploying.
+// The remainder is devoted to the install script.
+const CLOUD_DEPLOY_PROGRESS = 20;
+
+// Invokes f with an instance of the specified freedomjs module.
+// The module instance will be destroyed before the function resolves
+// or rejects. Intended for use with heavy-weight modules such as
+// those used for cloud.
+function oneShotModule_<T>(moduleName: string, f: (provider: any) => Promise<T>): Promise<T> {
+  try {
+    const m = freedom[moduleName]();
+    log.debug('created ' + moduleName + ' module');
+
+    const destructor = () => {
+      try {
+        freedom[moduleName].close(m);
+        log.debug('destroyed ' + moduleName + ' module');
+      } catch (e) {
+        log.debug('error destroying ' + moduleName + ' module: ' + e.message);
+      }
+    };
+
+    return f(m).then((result: T) => {
+      destructor();
+      return result;
+    }, (e: Error) => {
+      destructor();
+      throw e;
+    });
+  } catch (e) {
+    return Promise.reject(new Error('error creating ' + moduleName + ' module: ' + e.message));
+  }
+}
+
```

Comment:
If `f(m)` throws, we'll enter the `catch` block without calling `destructor()`. I think adding a `destructor()` call in the `catch` block before we return should do the trick.

(It'd be nice if the Promise API exposed a `.finally()` method, which we could use here to call `destructor()` in both the fulfilled and rejected case to eliminate some duplication. http://exploringjs.com/es6/ch_promises.html#_finally mentions adding this, but not sure if/when that will land, alas.)